File: solutions/scala/clock/1/src/main/scala/Clock.scala
case class Clock(hour: Int, minutes: Int) {
  def +(other: Clock): Clock = combineClocks(other, (a, b) => a + b)

  def -(other: Clock): Clock = combineClocks(other, (a, b) => a - b)

  private def combineClocks(other: Clock, operation: (Int, Int) => Int): Clock = {
    val totalMinutes = operation(this.minutes, other.minutes) + operation(this.hour, other.hour) * 60
    Clock(totalMinutes)
  }
}

object Clock {
  private val MinutesPerHour = 60
  private val HoursPerDay    = 24
  private val MinutesPerDay: Int = MinutesPerHour * HoursPerDay

  def apply(hour: Int, minutes: Int): Clock = {
    def totalMinutes = hour * MinutesPerHour + minutes
    def normalizedMinutes = ((totalMinutes % MinutesPerDay) + MinutesPerDay) % MinutesPerDay % MinutesPerHour
    def normalizedHour = ((totalMinutes % MinutesPerDay) + MinutesPerDay) % MinutesPerDay / MinutesPerHour

    new Clock(normalizedHour, normalizedMinutes)
  }

  def apply(minutes: Int): Clock = apply(0, minutes)
}

File: solutions/java/clock/1/src/main/java/Clock.java
class Clock {
    private int totalMinutes;

    Clock(int hours, int minutes) {
        this.totalMinutes = this.normalize(hours * 60L + minutes);
    }

    private static final int MINUTES_PER_DAY = 24 * 60;
    private static final int MINUTES_PER_HOUR = 60;
    private static final int SECONDS_PER_MINUTE = 60;

    void add(int minutes) {
        long updated = (long) this.totalMinutes + minutes;
        this.totalMinutes = this.normalize(updated);
    }

    @Override
    public String toString() {
        int hours = this.totalMinutes / MINUTES_PER_HOUR;
        int mins = this.totalMinutes % SECONDS_PER_MINUTE;
        return String.format("%02d:%02d", hours, mins);
    }

    @Override
    public boolean equals(Object object) {
        return this == object
               || object instanceof Clock other
                  && this.totalMinutes == other.totalMinutes;
    }

    private int normalize(long minutes) {
        long mod = minutes % MINUTES_PER_DAY;
        if (mod < 0) {
            mod += MINUTES_PER_DAY;
        }
        return (int) mod;
    }
}